// api-poc-client/example/app.js
var express = require('express');
var session = require('cookie-session');
var ApiClient = require('../lib/client-base');
var request = require('superagent');

// SETUP
var app = module.exports = express();
var client = new ApiClient({
    authServer: "http://sb1idtest:3500",
    clientId: "client-lib-test",
    clientSecret: "cefd35d8dbeeb1265e2dcfb5e779b1fe0c3591584f38e14446f139499d44848c7a51c409ebb6f81e116c8ad80d01370b68d0b3c0b41ba98257a294efaa1ba11a",
    scope: ["accountsRead","transactionsRead", "clientsRead","tagsRead"],
    redirectUri: "http://localhost:3560/auth/callback",
    debugLog: console.log
});
app.use(session({
    name: 'session-example-client',
    secret: 'little-elephant-crossing-borders'
}));

// AUTH

// ROUTES
app.get("/auth", client.authorize);

app.get("/auth/callback", client.authorizeCallback, function (req, res) {
    res.redirect("/");
})

app.get("/", client.secured, function (req, res, next) {
    var accessToken = req.session.access_token;
    request.get('http://sb1idtest:3500/api/v1/me')
        .set('Authorization', 'Bearer ' + accessToken)
        .end(function (err, result) {
            if (err) {
                return next(err);
            }
            res.send(result);
        });
})

app.use(function (err, req, res, next) {
    res.status(err.status || 500);
    res.send({
        message: err.message,
        error: err.filename + ":" + err.lineNumber
    });
})

// SERVER
var port = process.env.PORT || 3560;
var server = app.listen(port, function () {
    var host = server.address().address;
    console.log("Client lib example app running on " + host + ":" + port);
});
